core/src/main/java/com/yahoo/oak/NovaMemoryManager.java

```diff
@@ -0,0 +1,526 @@
+/*
+ * Copyright 2020, Verizon Media.
+ * Licensed under the terms of the Apache 2.0 license.
+ * Please see LICENSE file in the project root for terms.
+ */
+
+package com.yahoo.oak;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+
+class NovaMemoryManager implements MemoryManager {
+    static final int RELEASE_LIST_LIMIT = 124;
+    static final int CACHE_PADDING = 8;
+    static final int IDENTRY = 0;
+    static final int REFENTRY = 1;
+    
+    private static final NovaMMHeader HEADER =
+        new NovaMMHeader(); // for off-heap header operations
+    private static final int VERS_INIT_VALUE = 1;
+    private static final int OFF_HEAP_HEADER_SIZE = 8; /* Bytes */
+    private static final int VERSION_SIZE = 22;  // block offset get 42 bits on heap -  (including one bit for deleted)
+
+    
+    
+    private final ThreadIndexCalculator threadIndexCalculator;
+    private final List<List<SliceNova>> releaseLists;
+    private final AtomicInteger globalVersionNumber;
+    private final BlockMemoryAllocator allocator;
+
+    
+    private final AtomicLongArray tap;
```

Comment:
I am not sure what is better `AtomicLongArray` or normal array with UNSAFE.CAS where needed only...

Reply:
the use of atomicLongArrray is more correct in my opinion, 
since it guarantees the volatile guarantees, after entry is set every read after it being set sees this entry updated (with the new value).
Using just the unsafe CAS does not guarantee this,
we dont care about atomicity we care about preserving the order so we dont delete a slice that is being written to.
(If i recall correctly we there is not difference from performance point of view).